### PR TITLE
TINY-3785: Remove all uses of dom-globals

### PR DESCRIPTION
**Pull request by lnewson**
Related Ticket: TINY-3785

Description of Changes:

This removes dom-globals as a dependency. I still have to look at adding a lint rule to disable using global variables, etc... directly so that will be in the next PR. I also still have to rename `Option` to `Optional`, but trying to keep this just to removing dom-globals.

**Note:** This also fixes issues from having to upgrade to the TS 3.9 dom library.

Pre-checks:
* [x] ~Changelog entry added~
* [x] ~Tests have been added (if applicable)~
* [x] Branch prefixed with `feature/` for new features (if applicable)
* [x] ~License headers added on new files (if applicable)~

Review:
* [x] Milestone set
* [x] Review comments resolved

GitHub issues (if applicable):

### CODE REVIEW

File: modules/sugar/src/main/ts/ephox/sugar/api/search/Traverse.ts [hunk 1]
```diff
@@ -18,12 +17,15 @@ const owner = (element: SugarElement<Node>) => SugarElement.fromDom(element.dom(
...
 
 // The window element associated with the element
-const defaultView = (element: SugarElement<Node>) => SugarElement.fromDom(element.dom().ownerDocument.defaultView);
+const defaultView = (element: SugarElement<Node>) => SugarElement.fromDom(documentOrOwner(element).dom().defaultView);
 
-const parent = (element: SugarElement<Node>) => Option.from(element.dom().parentNode).map(SugarElement.fromDom);
+const parent = (element: SugarElement<Node>): Option<SugarElement<Node & ParentNode>> => Option.from(element.dom().parentNode).map(SugarElement.fromDom);
+
+// Cast down to just be SugarElement<Node>
+const parentNode = (element: SugarElement<Node>): Option<SugarElement<Node>> => parent(element) as any;
```
Comment (lnewson) [inline modules/sugar/src/main/ts/ephox/sugar/api/search/Traverse.ts:28]: New function here.

File: modules/tinymce/src/core/main/ts/SelectionOverrides.ts [hunk 2]
```diff
@@ -284,7 +283,8 @@ const SelectionOverrides = (editor: Editor): SelectionOverrides => {
           e.preventDefault();
           clipboardData.clearData();
           clipboardData.setData('text/html', realSelectionElement.outerHTML);
-          clipboardData.setData('text/plain', realSelectionElement.outerText);
+          // outerText is a nonstandard property and doesn't exist on Firefox, so fallback to innerText
+          clipboardData.setData('text/plain', (realSelectionElement as any).outerText || realSelectionElement.innerText);
```
Comment (lnewson) [inline modules/tinymce/src/core/main/ts/SelectionOverrides.ts:287]: Bug fixed here, as Firefox just returns undefined unlike all other browsers (the new types picked this up).

File: modules/alloy/src/main/ts/ephox/alloy/dragging/dragndrop/DropEvent.ts [hunk 3]
```diff
@@ -12,7 +11,8 @@ export interface DropEvent extends SimulatedEvent<EventArgs<DragEvent>> {
 
 export const createDropEventDetails = (config: DroppingConfig, event: SimulatedEvent<EventArgs<DragEvent>>): DropEvent => {
   const rawEvent = event.event().raw();
-  const transfer = rawEvent.dataTransfer;
+  // TODO: Should this handle if dataTransfer is null?
+  const transfer = rawEvent.dataTransfer as DataTransfer;
```
Comment (lnewson) [inline modules/alloy/src/main/ts/ephox/alloy/dragging/dragndrop/DropEvent.ts:15]: This was something the new types showed up as well. We weren't handling this previously either, so this is probably okay?

File: modules/imagetools/src/main/ts/ephox/imagetools/transformations/ImageResizerWebgl.ts [hunk 4]
```diff
@@ -189,17 +189,18 @@ function _createProgram(gl: WebGLRenderingContext): WebGLProgram {
   return program;
 }
 
-function _loadShader(gl: WebGLRenderingContext, source: string, type: string) {
-  const shader = gl.createShader((gl as any)[type]);
-  gl.shaderSource(shader, source);
-  gl.compileShader(shader);
+function _loadShader(gl: WebGLRenderingContext, source: string, type: string): Option<WebGLShader> {
+  return Option.from(gl.createShader((gl as any)[type])).map((shader) => {
```
Comment (lnewson) [inline modules/imagetools/src/main/ts/ephox/imagetools/transformations/ImageResizerWebgl.ts:193]: `gl.createShader` could have returned `null`, so I've converted this to an Option to handle the `null` case.

File: modules/imagetools/src/main/ts/ephox/imagetools/util/Conversions.ts [hunk 5]
```diff
@@ -173,7 +172,7 @@ function blobToDataUri(blob: Blob): Promise<string> {
     const reader = new FileReader();
 
     reader.onloadend = function () {
-      resolve(reader.result);
+      resolve(reader.result as string);
```
Comment (lnewson) [inline modules/imagetools/src/main/ts/ephox/imagetools/util/Conversions.ts:175]: `FileReader` can return either a `string` or an `ArrayBuffer`, depending on whether you call `readAsDataURL`, `readAsArrayBuffer`, etc... so we have to cast here unfortunately.

Comment (metricjs) [reply modules/imagetools/src/main/ts/ephox/imagetools/util/Conversions.ts:175]: could a code comment be helpful for this?

Comment (lnewson) [reply modules/imagetools/src/main/ts/ephox/imagetools/util/Conversions.ts:175]: It's done in quite a number of places and I really didn't want to have to carry that across in every place.

**Pull request merged**